samples/username-header/Core_6/Endpoint2/HandlerUsingCustomHeader.cs

```diff
@@ -11,9 +12,8 @@ public class HandlerUsingCustomHeader :
 
     public Task Handle(MyMessage message, IMessageHandlerContext context)
     {
-        var headers = context.MessageHeaders;
-        var usernameFromHeader = headers["UserName"];
-        log.Info($"Username extracted from header: {usernameFromHeader}");
+        var usernameFromThread = Thread.CurrentPrincipal.Identity.Name;
```

Comment:
`Thread.CurrentPrincipal` isn't guaranteed with async and shouldn't be used. Instead, `IPrincipal` should be resolved via DI by having an resolver for `IPrincipal` or by using a funky context extension like `.GetIdentity()`.

Reply:
The problem is like @ramonsmits pointed out that `Thread.CurrentPrincipal` behavior is different depending on the runtime. For the .NET Framework floating the principle over the ExecutionContext works. For .NET Core 3 it works as well because the behavior has been aligned as outline in https://github.com/dotnet/corefx/issues/34489#issuecomment-454860866. For pre 3.0 .NET Core the behavior is to not float the principal into the execution context. So if you execute the following code under .NET Core 3.0 

```
            var identity = new ClaimsIdentity();
            identity.AddClaim(new Claim(ClaimTypes.Name, "User1"));
            Thread.CurrentPrincipal = new ClaimsPrincipal(identity);

            Console.WriteLine($"Current thread: {Thread.CurrentThread.ManagedThreadId}");
            // Check the current user
            Console.WriteLine($"Current user: {Thread.CurrentPrincipal?.Identity.Name}");

            // For the method to complete asynchronously
            await Task.Yield();

            Console.WriteLine($"Current thread: {Thread.CurrentThread.ManagedThreadId}");
            // Check the current user after
            Console.WriteLine($"Current user: {Thread.CurrentPrincipal?.Identity.Name}");
```

the result is

```
Current thread: 1
Current user: User1
Current thread: 4
Current user: User1
```

if executed under .NET Core 2.2 the result is

```
Current thread: 1
Current user: User1
Current thread: 4
Current user:
```

Reply:
This is one of the reasons why the migration guide here mentions to avoid `Thread.CurrentPrincipal` or `ClaimsPrincipal.Current`

https://docs.microsoft.com/en-us/aspnet/core/migration/claimsprincipal-current

and the guidance has four ways of doing it

https://docs.microsoft.com/en-us/aspnet/core/migration/claimsprincipal-current?view=aspnetcore-3.0#retrieve-the-current-user-in-an-aspnet-core-app

Reply:
I will send it a very rough proposal to discuss